lively.ide/studio/asset-browser.cp.js

```diff
@@ -0,0 +1,918 @@
+    return {
+      container: {},
+      allowDraggingAssets: {
+        defaultValue: false
+      }
+    };
+  }
+
+  get assetBrowserAsPopup () {
+    return !this.view.ownerChain().some(m => m.isWindow);
+  }
+
+  get expose () {
+    return ['activate', 'container', 'close', 'initialize', 'block'];
+  }
+
+  get bindings () {
+    return [
+      { target: 'upload button', signal: 'onMouseDown', handler: 'openFilePicker' },
+      // See https://github.com/LivelyKernel/lively.next/issues/1057
+      { target: 'delete button', signal: 'onMouseUp', handler: 'deleteAsset' },
+      { target: 'selection button', signal: 'onMouseDown', handler: 'confirm' },
+      { target: 'search input', signal: 'inputChanged', handler: 'filterAssets' },
+      {
+        target: 'search clear button',
+        signal: 'onMouseDown',
+        handler: () => {
+          this.ui.searchInput.textString = '';
+          this.filterAssets();
+        }
+      },
+      {
+        target: 'search input',
+        signal: 'onKeyDown',
+        handler: (evt) => {
+          if (evt.key === 'Escape') {
+            this.ui.searchInput.textString = '';
+            this.filterAssets();
+          }
+        }
+      }
+    ];
+  }
+
+  filterAssets () {
+    const needle = this.ui.searchInput.textString.toLowerCase();
+    this.ui.assets.submorphs.forEach(s => {
+      if (!s.isAssetPreview) return;
+      s.visible = s.isLayoutable = true;
+    });
+    if (needle.trim() === '') {
+      this.ui.noResultsIndicator.visible = false;
+      this.ui.searchClearButton.visible = false;
+      return;
+    }
+
+    this.ui.searchClearButton.visible = true;
+
+    this.ui.assets.submorphs.forEach(s => {
+      if (!s.isAssetPreview) return;
+      if (!s.assetName.toLowerCase().includes(needle)) s.visible = s.isLayoutable = false;
+    });
+    if (this.ui.assets.submorphs.every(s => !s.isAssetPreview || !s.visible)) this.ui.noResultsIndicator.visible = true;
+    else this.ui.noResultsIndicator.visible = false;
+  }
+
+  confirm () {
+    // _promise is initialized inside of the popup, but not the window
+    if (this._promise) {
+      this._promise.resolve(this.selectedAsset.imageUrl);
+      this.container.remove();
+    } else {
+      const image = new Image({ imageUrl: this.selectedAsset.imageUrl });
+      image.determineNaturalExtent().then((extent) => {
+        const maxWidth = 500;
+        const maxHeight = 500;
+        const scaleFactor = Math.min(maxWidth / extent.x, maxHeight / extent.y);
+        image.extent = extent.scaleBy(scaleFactor);
+        image.openInWorld();
+      });
+    }
+  }
+
+  async initialize () {
+    this.ui.assetTypeSelector.enabled = false;
+    this.view.visible = false;
+    const li = $world.showLoadingIndicatorFor(null, 'Enumerating project assets...');
+    await this.listAssets();
+    this.ui.selectionButton.disable();
+    this.view.visible = true;
+    li.remove();
+  }
+
+  block () {
+    const fader = morph({
```

Comment:
We now have at least 2 blocking faders in the world, another one being the one we use for the window switcher. Can we instead centralise this logic in the world?

Reply:
For the window switcher, the blurring fader is part of the `WindowSwitcher` itself. Therefore, I would keep it this way for now, okay?